README.md
SOAP-WSSE
=========

Simply library to sign and verify SOAP XML requests using the
BinarySecurityToken specification.

Improvements welcome :-)



Continuous integration status:

.. image:: https://travis-ci.org/mvantellingen/py-soap-wsse.svg?branch=master
    :target: https://travis-ci.org/mvantellingen/py-soap-wsse

.. image:: https://coveralls.io/repos/mvantellingen/py-soap-wsse/badge.png?branch=master
    :alt: Coverage
    :target: https://coveralls.io/r/mvantellingen/py-soap-wsse


References
----------
 - http://vsza.hu/thesis-beta.pdf
 - https://github.com/dnet/SudsSigner


CHANGES
0.2.1 (2015-06-04)
==================
 - Don't try to validate empty responses in the Suds plugin
 - Remove warning log message in `signing.verify_envelope`


0.2.0 (26-02-2015)
==================

Remove suds as a dependency and only check for its existence in the plugin
module.


0.1.2  (18-02-2015)
===================
 - Fix import errors
 - Add extra unittests 


0.1.1  (18-02-2015)
===================

 - Use MIT license


0.1 (17-02-2015)
================

Initial release


tests/test_soap_wsse.py
import os
import soap_wsse


KEY_FILE = os.path.join(
    os.path.dirname(os.path.realpath(__file__)), 'soap_wsse_keys.pem')


XML="""
<soapenv:Envelope xmlns:mvt="http://github.com/mvantellingen"
    xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/"
    xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
    xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/">
  <soapenv:Header></soapenv:Header>
  <soapenv:Body>
    <mvt:Function>
      <mvt:Argument>OK</mvt:Argument>
    </mvt:Function>
  </soapenv:Body>
</soapenv:Envelope>
""".strip()


def test_sign():
    signed_xml = soap_wsse.sign_envelope(XML, KEY_FILE)
    result = soap_wsse.verify_envelope(signed_xml, KEY_FILE)
    assert result is True


def test_sign_failed():
    signed_xml = soap_wsse.sign_envelope(XML, KEY_FILE)
    signed_xml = signed_xml.replace('OK', 'NOT OK!')

    result = soap_wsse.verify_envelope(signed_xml, KEY_FILE)
    assert result is False


tests/test_suds_plugin.py
import os.path

import pytest
from pretend import stub

from soap_wsse import CertificationError
from soap_wsse import suds_plugin


KEY_FILE = os.path.join(
    os.path.dirname(os.path.realpath(__file__)), 'soap_wsse_keys.pem')


XML="""
<soapenv:Envelope xmlns:mvt="http://github.com/mvantellingen"
    xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/"
    xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
    xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/">
  <soapenv:Header></soapenv:Header>
  <soapenv:Body>
    <mvt:Function>
      <mvt:Argument>OK</mvt:Argument>
    </mvt:Function>
  </soapenv:Body>
</soapenv:Envelope>
""".strip()

XML_SIGNED = """
<soapenv:Envelope xmlns:mvt="http://github.com/mvantellingen" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/">
   <soapenv:Header>
      <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
         <wsse:BinarySecurityToken wsu:Id="id-e173afc8-2e7f-4d37-9ef5-160b84e210ab" EncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary" ValueType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3">MIIEgzCCA2ugAwIBAgIJAMkiZttOMvKxMA0GCSqGSIb3DQEBBQUAMIGHMQswCQYDVQQGEwJOTDETMBEGA1UECBMKR2VsZGVybGFuZDEMMAoGA1UEBxMDRWRlMRAwDgYDVQQKEwdMdWtraWVuMRgwFgYDVQQDEw9wb24ubHVra2llbi5jb20xKTAnBgkqhkiG9w0BCQEWGm0udmFudGVsbGluZ2VuQGx1a2tpZW4uY29tMB4XDTE1MDEyOTE2MDk0OVoXDTE3MTAyNTE2MDk0OVowgYcxCzAJBgNVBAYTAk5MMRMwEQYDVQQIEwpHZWxkZXJsYW5kMQwwCgYDVQQHEwNFZGUxEDAOBgNVBAoTB0x1a2tpZW4xGDAWBgNVBAMTD3Bvbi5sdWtraWVuLmNvbTEpMCcGCSqGSIb3DQEJARYabS52YW50ZWxsaW5nZW5AbHVra2llbi5jb20wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDC/R9s5eApMoc+B9G9IklGtcjkOAftHmntTrbKGACz9dTRIURCeqtuJuEvvvpAarnFftIb39tAvdrwkZIkmaAUowcD7OHsOJlsojNY7vYxju8Clx45XJ9HcvL6EkHpd2lRuLx9dSfwTfGKtnTaFhJt8ZnLjNSbWK/5IHHvrv/gYtDKhahK6ncBW2k5nXZf6Wn2Rn/RpjkSoQL12Gmyh47EXMvbA9HpxxqDlPWBBNp6hpGhGOkg0EVvRgnKwzkqvTyB6571LsNMVe7U+gkmd9GHxx8t3cloWLG9RSD/Qr0ahQpFPI00c4dHN0a1LG4WCzbFN8mZYr3WSqsx1OUY1eHJAgMBAAGjge8wgewwHQYDVR0OBBYEFKicd7yRgkGBJJoO/s49iSF3N91CMIG8BgNVHSMEgbQwgbGAFKicd7yRgkGBJJoO/s49iSF3N91CoYGNpIGKMIGHMQswCQYDVQQGEwJOTDETMBEGA1UECBMKR2VsZGVybGFuZDEMMAoGA1UEBxMDRWRlMRAwDgYDVQQKEwdMdWtraWVuMRgwFgYDVQQDEw9wb24ubHVra2llbi5jb20xKTAnBgkqhkiG9w0BCQEWGm0udmFudGVsbGluZ2VuQGx1a2tpZW4uY29tggkAySJm204y8rEwDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQUFAAOCAQEAo43YKvwGepefY3mazx+PUa5OCozHHNtvpZpXRtN/3bwggXZdJqTyJjmlEQBZz/yAyJL5Ar8FtMenR4Ki8E9Esn09L/l2rA0JvLP8IMBZHfqdDM2Za5zJwp541y4jRjcNlVJ57bwby3DlBp8u70wrtGp8PNC1cGLr9Wj61mERjQAeIn4Qv8JBKiuvQ+YiHN5x1baOCxOWYGFlukXiGDcnNse0BC144yBJraoLzCI6VQYSws1n33VyjPPySpDeR2/JdS6ZO1E/yhuqqwXt8p3g3i7brkpxbQYYeBQA0idDBO7bVQOBoXYGbIN2AIrkPnm4zkki1kivh5NP4PakO6TPDQ==</wsse:BinarySecurityToken>
         <Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
            <SignedInfo>
               <CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
               <SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
               <Reference URI="#id-b88aa1ed-1b06-4b40-b110-5c239a6b5ce3">
                  <Transforms>
                     <Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#">
                        <ec:InclusiveNamespaces PrefixList="urn" xmlns:ec="http://www.w3.org/2001/10/xml-exc-c14n#"/>
                     </Transform>
                  </Transforms>
                  <DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
                  <DigestValue>NRPhamBfgxgTWPf6+8Xzga+YMtg=</DigestValue>
               </Reference>
            </SignedInfo>
            <SignatureValue>ms8rrvFl12PTyhZFBBI6l5T9wCCljGamaiVzGS3HcTGw+gQ5YwHSPSuwAAxhSXKi
krHVN0zEw6p0HZC5RJR/XFrCLPScsrDQOjIP2pKU3uuKR7wFLFDzygS5yU7qpuSn
TdoRuqyOgT5lgvDlDBXn534cQHW3yOUVUl3u+QWdrDtVehag17JBDA89db5KiOfW
ReaTUxHjYmYJaEtY4HLr9PGvtS4pEpr+FCvHoq0aKI40BqyGar4G1/8tavWGaFV0
vMSZrBGvRxqq6Gotjjt47LhlYq3JvpLLLi+SOjF5388LMKZ+gLLapIY0OhvnGEvl
JvsAu49CHgPAlLdF3wvIgA==</SignatureValue>
            <KeyInfo>
               <wsse:SecurityTokenReference wsse:TokenType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3">
                  <wsse:Reference ValueType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3" URI="#id-e173afc8-2e7f-4d37-9ef5-160b84e210ab"/>
               </wsse:SecurityTokenReference>
            </KeyInfo>
         </Signature>
      </wsse:Security>
   </soapenv:Header>
   <soapenv:Body ns0:Id="id-b88aa1ed-1b06-4b40-b110-5c239a6b5ce3" xmlns:ns0="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
      <mvt:Function>
         <mvt:Argument>OK</mvt:Argument>
      </mvt:Function>
   </soapenv:Body>
</soapenv:Envelope>
"""


def test_plugin_sending():
    plugin = suds_plugin.WssePlugin(KEY_FILE)

    context = stub(envelope=XML)
    plugin.sending(context)

def test_plugin_received():
    plugin = suds_plugin.WssePlugin(KEY_FILE)

    context = stub(reply=XML_SIGNED)
    with pytest.raises(CertificationError):
        plugin.received(context)


def test_plugin_received_without_signature():
    plugin = suds_plugin.WssePlugin(KEY_FILE)

    context = stub(reply=XML)
    with pytest.raises(CertificationError):
        plugin.received(context)


src/soap_wsse/signing.py
"""
    soap_wsse.signing
    ~~~~~~~~~~~~~~~~~

    Library to sign SOAP requests with WSSE tokens.

"""
import logging
import base64
import os
from uuid import uuid4

import dm.xmlsec.binding as xmlsec
from dm.xmlsec.binding.tmpl import Signature
from lxml import etree
from OpenSSL import crypto

from soap_wsse import ns


logger = logging.getLogger(__name__)


BODY_XPATH = etree.XPath(
    '/SOAP-ENV:Envelope/SOAP-ENV:Body', namespaces=ns.NSMAP)
HEADER_XPATH = etree.XPath(
    '/SOAP-ENV:Envelope/SOAP-ENV:Header', namespaces=ns.NSMAP)
SECURITY_XPATH = etree.XPath('wsse:Security', namespaces=ns.NSMAP)
TIMESTAMP_XPATH = etree.XPath('wsu:Timestamp', namespaces=ns.NSMAP)

C14N = 'http://www.w3.org/2001/10/xml-exc-c14n#'
XMLDSIG_SHA1 = 'http://www.w3.org/2000/09/xmldsig#sha1'


def ns_id(tagname, suds_ns):
    return '{{{0}}}{1}'.format(suds_ns[1], tagname)

WSU_ID = ns_id('Id', ns.wsuns)
BINARY_TOKEN_TYPE = (
    'http://docs.oasis-open.org/wss/2004/01/' +
    'oasis-200401-wss-x509-token-profile-1.0#X509v3')


def log_errors(filename, line, func, errorObject, errorSubject, reason, msg):
    info = []
    if errorObject != 'unknown':
        info.append('obj=' + errorObject)
    if errorSubject != 'unknown':
        info.append('subject=' + errorSubject)
    if msg.strip():
        info.append('msg=' + msg)
    if info:
        logger.debug('%s:%d(%s)' % (filename, line, func), ' '.join(info))


class CertificationError(Exception):
    pass


# Initialize the xmlsec library
xmlsec.initialize()
xmlsec.set_error_callback(log_errors)


class SignQueue(object):
    WSU_ID = ns_id('Id', ns.wsuns)
    DS_DIGEST_VALUE = ns_id('DigestValue', ns.dsns)
    DS_REFERENCE = ns_id('Reference', ns.dsns)
    DS_TRANSFORMS = ns_id('Transforms', ns.dsns)

    def __init__(self):
        self.queue = []

    def push_and_mark(self, element):
        unique_id = get_unique_id()
        element.set(self.WSU_ID, unique_id)
        self.queue.append(unique_id)

    def insert_references(self, signature):
        signed_info = signature.find('ds:SignedInfo', namespaces=ns.NSMAP)
        nsmap = {ns.ecns[0]: ns.ecns[1]}

        for element_id in self.queue:
            reference = etree.SubElement(
                signed_info, self.DS_REFERENCE,
                {'URI': '#{0}'.format(element_id)})
            transforms = etree.SubElement(reference, self.DS_TRANSFORMS)
            node = set_algorithm(transforms, 'Transform', C14N)

            elm = _create_element(node, 'ec:InclusiveNamespaces', nsmap)
            elm.set('PrefixList', 'urn')

            set_algorithm(reference, 'DigestMethod', XMLDSIG_SHA1)
            etree.SubElement(reference, self.DS_DIGEST_VALUE)


def sign_envelope(envelope, key_file):
    """Sign the given soap request with the given key"""
    doc = etree.fromstring(envelope)
    body = get_body(doc)

    queue = SignQueue()
    queue.push_and_mark(body)

    security_node = ensure_security_header(doc, queue)
    security_token_node = create_binary_security_token(key_file)
    signature_node = Signature(
        xmlsec.TransformExclC14N, xmlsec.TransformRsaSha1)

    security_node.append(security_token_node)
    security_node.append(signature_node)
    queue.insert_references(signature_node)

    key_info = create_key_info_node(security_token_node)
    signature_node.append(key_info)

    # Sign the generated xml
    xmlsec.addIDs(doc, ['Id'])
    dsigCtx = xmlsec.DSigCtx()
    dsigCtx.signKey = xmlsec.Key.load(key_file, xmlsec.KeyDataFormatPem, None)
    dsigCtx.sign(signature_node)
    return etree.tostring(doc)


def verify_envelope(reply, key_file):
    """Verify that the given soap request is signed with the certificate"""
    doc = etree.fromstring(reply)
    node = doc.find(".//{%s}Signature" % xmlsec.DSigNs)
    if node is None:
        raise CertificationError("No signature node found")
    dsigCtx = xmlsec.DSigCtx()

    xmlsec.addIDs(doc, ['Id'])
    signKey = xmlsec.Key.load(key_file, xmlsec.KeyDataFormatPem)
    signKey.name = os.path.basename(key_file)

    dsigCtx.signKey = signKey
    try:
        dsigCtx.verify(node)
    except xmlsec.VerificationError:
        return False
    return True


def get_unique_id():
    return 'id-{0}'.format(uuid4())


def set_algorithm(parent, name, value):
    return etree.SubElement(parent, ns_id(name, ns.dsns), {'Algorithm': value})


def get_body(envelope):
    (body,) = BODY_XPATH(envelope)
    return body


def create_key_info_node(security_token):
    """Create the KeyInfo node for WSSE.

    Note that this currently only supports BinarySecurityTokens

    Example of the generated XML:

        <ds:KeyInfo Id="KI-24C56C5B3448F4BE9D141094243396829">
            <wsse:SecurityTokenReference
                wsse11:TokenType="{{ BINARY_TOKEN_TYPE }}">
                <wsse:Reference
                    URI="#X509-24C56C5B3448F4BE9D141094243396828"
                    ValueType="{{ BINARY_TOKEN_TYPE }}"/>
           </wsse:SecurityTokenReference>
        </ds:KeyInfo>

    """
    key_info = etree.Element(ns_id('KeyInfo', ns.dsns))

    sec_token_ref = etree.SubElement(
        key_info, ns_id('SecurityTokenReference', ns.wssens))
    sec_token_ref.set(
        ns_id('TokenType', ns.wssens), security_token.get('ValueType'))

    reference = etree.SubElement(sec_token_ref, ns_id('Reference', ns.wssens))
    reference.set('ValueType', security_token.get('ValueType'))
    reference.set('URI', '#%s' % security_token.get(WSU_ID))
    return key_info


def create_binary_security_token(key_file):
    """Create the BinarySecurityToken node containing the x509 certificate.

    """
    node = etree.Element(
        ns_id('BinarySecurityToken', ns.wssens),
        nsmap={ns.wssens[0]: ns.wssens[1]})
    node.set(ns_id('Id', ns.wsuns), get_unique_id())
    node.set('EncodingType', ns.wssns[1] + 'Base64Binary')
    node.set('ValueType', BINARY_TOKEN_TYPE)

    with open(key_file) as fh:
        cert = crypto.load_certificate(crypto.FILETYPE_PEM, fh.read())
        node.text = base64.b64encode(
            crypto.dump_certificate(crypto.FILETYPE_ASN1, cert))
    return node


def ensure_security_header(envelope, queue):
    """Insert a security XML node if it doesn't exist otherwise update it.

    """
    (header,) = HEADER_XPATH(envelope)
    security = SECURITY_XPATH(header)
    if security:
        for timestamp in TIMESTAMP_XPATH(security[0]):
            queue.push_and_mark(timestamp)
        return security[0]
    else:
        nsmap = {
            'wsu': ns.wsuns[1],
            'wsse': ns.wssens[1],
        }
        return _create_element(header, 'wsse:Security', nsmap)


def _create_element(parent, name, nsmap):
    prefix, name = name.split(':', 1)
    tag_name = '{%s}%s' % (nsmap[prefix], name)

    if parent is not None:
        return etree.SubElement(parent, tag_name, nsmap=nsmap)
    else:
        return etree.Element(tag_name, nsmap=nsmap)
